### PR TITLE
feat: [PL-43723]: Create batch API to fetch licenses for multiple accounts

### PR DESCRIPTION
**Pull request by AyushAgrawal250**
[Problem]

A batch API is required to fetch licenses of multiple accounts in single call.

[Solution]

Added an API which accepts multiple account ids and will return the licenses associated to those account ids in a single call.

[Test]

Tested with local setup

Jira: https://harness.atlassian.net/browse/PL-43723

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/57034)

### CODE REVIEW

File: 960-ng-license-beans/src/main/java/io/harness/licensing/beans/modules/BatchModuleLicenseRequestDTO.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.List;
+
+@OwnedBy(HarnessTeam.PL)
+@Data
+@Builder
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@Schema(name = "BatchModuleLicenseRequest", description = "This contains the data required for batch call.")
+public class BatchModuleLicenseRequestDTO {
+    @Schema(description = "Specifies the list of account Id.") @NotNull List<String> accountIds;
```
Comment (niketan-nath) [inline 960-ng-license-beans/src/main/java/io/harness/licensing/beans/modules/BatchModuleLicenseRequestDTO.java:20]:
```suggestion
    @Schema(description = "Specifies the list of account ids.") @NotNull List<String> accountIds;
```

**Review by niketan-nath — APPROVED** [review]
LGTM

**Comment by niketan-nath** [comment]
trigger uts

**Comment by niketan-nath** [comment]
trigger copyrightcheck

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57034)* status: 2 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/resources/AdminAccountResource.java` line 349 at r2](https://reviewable.io/reviews/harness/harness-core/57034#-NmFiQt80XCROs28c0Wo:-NmFiQt80XCROs28c0Wp:boejngp) ([raw file](https://github.com/harness/harness-core/blob/a8c0c49024cc35f4d11556c779f46b90c4cfe0d3/400-rest/src/main/java/software/wings/resources/AdminAccountResource.java#L349)):*
> ```Java
> 
>   @POST
>   @Path("ng/license/batch")
> ```

Can we keep the endpoint as just /license like how we have /summary endpoint
___
*[`400-rest/src/main/java/software/wings/resources/AdminAccountResource.java` line 350 at r2](https://reviewable.io/reviews/harness/harness-core/57034#-NmFifFF0guJaFK6cbrc:-NmFifFF0guJaFK6cbrd:b-4boyr5) ([raw file](https://github.com/harness/harness-core/blob/a8c0c49024cc35f4d11556c779f46b90c4cfe0d3/400-rest/src/main/java/software/wings/resources/AdminAccountResource.java#L350)):*
> ```Java
>   @POST
>   @Path("ng/license/batch")
>   public RestResponse<List<AccountLicenseDTO>> getBatchOfNgAccountLicenses(@Body BatchModuleLicenseRequestDTO batchModuleLicenseRequest) {
> ```

AccountsListRequestDTO instead of BatchModuleLicenseRequestDTO
___
*[`960-ng-license-beans/src/main/java/io/harness/licensing/beans/modules/BatchModuleLicenseRequestDTO.java` line 25 at r2](https://reviewable.io/reviews/harness/harness-core/57034#-NmFj7_o2-pY9cVS0mEY:-NmFj11e5P_9kcK55exE:b-frtsps) ([raw file](https://github.com/harness/harness-core/blob/a8c0c49024cc35f4d11556c779f46b90c4cfe0d3/960-ng-license-beans/src/main/java/io/harness/licensing/beans/modules/BatchModuleLicenseRequestDTO.java#L25)):*
> ```Java
> @FieldDefaults(level = AccessLevel.PRIVATE)
> @JsonIgnoreProperties(ignoreUnknown = true)
> @Schema(name = "BatchModuleLicenseRequest", description = "This contains the data required for batch call.")
> ```

"This contains the payload required for retrieve licenses of multiple accounts at once

**Comment by AyushAgrawal250** [comment]
> > Reviewed all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/57034)_ status: 2 file(s) to review. 1 change requests. 0 of 2 approvals obtained.
> 
> _[`400-rest/src/main/java/software/wings/resources/AdminAccountResource.java` line 349 at r2](https://reviewable.io/reviews/harness/harness-core/57034#-NmFiQt80XCROs28c0Wo:-NmFiQt80XCROs28c0Wp:boejngp) ([raw file](https://github.com/harness/harness-core/blob/a8c0c49024cc35f4d11556c779f46b90c4cfe0d3/400-rest/src/main/java/software/wings/resources/AdminAccountResource.java#L349)):_
> 
> > ```java
> >   @POST
> >   @Path("ng/license/batch")
> > ```
> 
> Can we keep the endpoint as just /license like how we have /summary endpoint
> 
> _[`400-rest/src/main/java/software/wings/resources/AdminAccountResource.java` line 350 at r2](https://reviewable.io/reviews/harness/harness-core/57034#-NmFifFF0guJaFK6cbrc:-NmFifFF0guJaFK6cbrd:b-4boyr5) ([raw file](https://github.com/harness/harness-core/blob/a8c0c49024cc35f4d11556c779f46b90c4cfe0d3/400-rest/src/main/java/software/wings/resources/AdminAccountResource.java#L350)):_
> 
> > ```java
> >   @POST
> >   @Path("ng/license/batch")
> >   public RestResponse<List<AccountLicenseDTO>> getBatchOfNgAccountLicenses(@Body BatchModuleLicenseRequestDTO batchModuleLicenseRequest) {
> > ```
> 
> AccountsListRequestDTO instead of BatchModuleLicenseRequestDTO
> 
> _[`960-ng-license-beans/src/main/java/io/harness/licensing/beans/modules/BatchModuleLicenseRequestDTO.java` line 25 at r2](https://reviewable.io/reviews/harness/harness-core/57034#-NmFj7_o2-pY9cVS0mEY:-NmFj11e5P_9kcK55exE:b-frtsps) ([raw file](https://github.com/harness/harness-core/blob/a8c0c49024cc35f4d11556c779f46b90c4cfe0d3/960-ng-license-beans/src/main/java/io/harness/licensing/beans/modules/BatchModuleLicenseRequestDTO.java#L25)):_
> 
> > ```java
> > @FieldDefaults(level = AccessLevel.PRIVATE)
> > @JsonIgnoreProperties(ignoreUnknown = true)
> > @Schema(name = "BatchModuleLicenseRequest", description = "This contains the data required for batch call.")
> > ```
> 
> "This contains the payload required for retrieve licenses of multiple accounts at once

I checked with `/licenses` as the endpoint and it's not conflicting with any other endpoints so I am updating it to `/licenses`.

`AccountLicenseListRequestDTO` makes more sense.

I will update the description.

**Review by Xin-Harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57034)* status: 2 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/resources/AdminAccountResource.java` line 106 at r3](https://reviewable.io/reviews/harness/harness-core/57034#-NmGwqvW5h3_9Wk0v9FW:-NmGwqvW5h3_9Wk0v9FX:b-a2st8q) ([raw file](https://github.com/harness/harness-core/blob/63e8c8ddf626a653a394812f9b6dc4d035e9035d/400-rest/src/main/java/software/wings/resources/AdminAccountResource.java#L106)):*
> ```Java
>   }
> 
>   @POST
> ```

400-rest/AdminAccountResource is not the right place to add this API because it doesn't have to go to CG and then use adminLicense client again to go to NG to retrieve license info. 
It should directly added to 930-ng-license-manager/licenseResource, please take a look and you would understand it.
In addition, I would recommend not use license as end point as it is very vague to map with a batch with licenses.

**Comment by AyushAgrawal250** [comment]
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/57034)_ status: 2 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.
> 
> _[`400-rest/src/main/java/software/wings/resources/AdminAccountResource.java` line 106 at r3](https://reviewable.io/reviews/harness/harness-core/57034#-NmGwqvW5h3_9Wk0v9FW:-NmGwqvW5h3_9Wk0v9FX:b-a2st8q) ([raw file](https://github.com/harness/harness-core/blob/63e8c8ddf626a653a394812f9b6dc4d035e9035d/400-rest/src/main/java/software/wings/resources/AdminAccountResource.java#L106)):_
> 
> > ```java
> >   }
> > 
> >   @POST
> > ```
> 
> 400-rest/AdminAccountResource is not the right place to add this API because it doesn't have to go to CG and then use adminLicense client again to go to NG to retrieve license info. It should directly added to 930-ng-license-manager/licenseResource, please take a look and you would understand it. In addition, I would recommend not use license as end point as it is very vague to map with a batch with licenses.

All the license APIs for Admin Portal are there in the AdminAccountResource, so I have added one more there, the overhead of shifting things to NG Manager will be more as then we need to shift all the license APIs that are already present. Also, these calls are made in the background to fetch module licenses so latency shouldn’t be that big of an issue as well.
cc: @niketan-nath @ankit-harness

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57034#-:-NnDoMtDBjWpCdTmVbpL:bnfp4nl)* status: Some of the checks are too old. 2 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by niketan-nath — APPROVED** [review]

**Comment by niketan-nath** [comment]
trigger uts

**Comment by AyushAgrawal250** [comment]
trigger ut9

**Comment by AyushAgrawal250** [comment]
API Response:

```
➜  ~ curl --location 'http://localdev.harness.io/api/admin/accounts/license' \
--header 'Authorization: AdminPortal eyJhbGciOiJIUzI1NiJ9.eyJqdGkiOiIzOWFlMjZmMC03ZDBkLTRjZTItYjdhZS0yZjVjMWZkN2I5ZGEiLCJpYXQiOjE3MDQzODA1OTEsImV4cCI6MTcwNDM4NDE5MSwiaXNzIjoiSGFybmVzcyBJbmMifQ.FC1HMN6GSAojnfVwwY37otWFiDqSRvvaAhE832HuW1E' \
--header 'Accept: application/json' \
--header 'Content-Type: application/json' \
--data '{
    "accountIds" : [
        "kmpySmUISimoRrJL6NL73w",
        "kmpySmUISimoRrJL6NL73w"
    ]
}'
{"metaData":{},"resource":[{"accountId":"kmpySmUISimoRrJL6NL73w","allModuleLicenses":{"SEI":[{"id":"657aa9b1d7a57577a29d539c","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"SEI","edition":"FREE","status":"ACTIVE","startTime":1701062158329,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdBy":{"uuid":null,"name":null,"email":"","externalUserId":null},"lastModifiedAt":1703157184910,"numberOfContributors":687}],"CHAOS":[],"CET":[],"CD":[{"id":"657b564983511c3c8d377193","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CD","edition":"FREE","status":"ACTIVE","startTime":1702581833799,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833801,"lastModifiedAt":1702581833801,"cdLicenseType":"SERVICES","workloads":5}],"CV":[],"STO":[],"CE":[{"id":"657b564983511c3c8d377195","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CE","edition":"FREE","status":"ACTIVE","startTime":1702581833808,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833809,"lastModifiedAt":1702581833809,"spendLimit":250000}],"CI":[{"id":"657b564983511c3c8d377194","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CI","edition":"FREE","status":"ACTIVE","startTime":1702581833804,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833805,"lastModifiedAt":1702581833805,"numberOfCommitters":-1}],"SRM":[{"id":"657b564983511c3c8d377197","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"SRM","edition":"FREE","status":"ACTIVE","startTime":1702581833814,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833816,"lastModifiedAt":1702581833816,"numberOfServices":5}],"IDP":[],"CF":[{"id":"657b564983511c3c8d377196","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CF","edition":"FREE","status":"ACTIVE","startTime":1702581833811,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833812,"lastModifiedAt":1702581833812,"numberOfUsers":2,"numberOfClientMAUs":25000}],"IACM":[]}},{"accountId":"kmpySmUISimoRrJL6NL73w","allModuleLicenses":{"SEI":[{"id":"657aa9b1d7a57577a29d539c","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"SEI","edition":"FREE","status":"ACTIVE","startTime":1701062158329,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdBy":{"uuid":null,"name":null,"email":"","externalUserId":null},"lastModifiedAt":1703157184910,"numberOfContributors":687}],"CHAOS":[],"CET":[],"CD":[{"id":"657b564983511c3c8d377193","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CD","edition":"FREE","status":"ACTIVE","startTime":1702581833799,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833801,"lastModifiedAt":1702581833801,"cdLicenseType":"SERVICES","workloads":5}],"CV":[],"STO":[],"CE":[{"id":"657b564983511c3c8d377195","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CE","edition":"FREE","status":"ACTIVE","startTime":1702581833808,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833809,"lastModifiedAt":1702581833809,"spendLimit":250000}],"CI":[{"id":"657b564983511c3c8d377194","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CI","edition":"FREE","status":"ACTIVE","startTime":1702581833804,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833805,"lastModifiedAt":1702581833805,"numberOfCommitters":-1}],"SRM":[{"id":"657b564983511c3c8d377197","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"SRM","edition":"FREE","status":"ACTIVE","startTime":1702581833814,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833816,"lastModifiedAt":1702581833816,"numberOfServices":5}],"IDP":[],"CF":[{"id":"657b564983511c3c8d377196","accountIdentifier":"kmpySmUISimoRrJL6NL73w","moduleType":"CF","edition":"FREE","status":"ACTIVE","startTime":1702581833811,"expiryTime":9223372036854775807,"premiumSupport":false,"selfService":true,"createdAt":1702581833812,"lastModifiedAt":1702581833812,"numberOfUsers":2,"numberOfClientMAUs":25000}],"IACM":[]}}],"responseMessages":[]}%
➜  ~
```

**Review by Xin-Harness — APPROVED** [review]
> Reviewed 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/57034)* status: 3 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Pull request merged**